### PR TITLE
quadlet: Add a network requirement on .image and .containers units

### PR DESCRIPTION
**Pull request by jbtrystram**
If a container unit starts on boot with a dependency on `default.target`
the image unit may start too soon, before network is ready. This cause
the unit to fail to pull the image.
Add a dependency on `network-online.target` to make sure image pulls
don't fail.

#### Does this PR introduce a user-facing change?
yes 

```release-note
Image oneshot units generated through quadlet will now have a dependency on `network-online.target`
```

Fixes https://github.com/containers/podman/issues/21873

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
@ygalblum @alexlarsson @mheon PTAL

**Comment by ygalblum** [comment]
The specific code LGTM. But, it addresses only the `.image` files. Doesn't the same rational apply for `.container` and `.kube` files?

In addition, while the documentation states that setting `After=` will remove this dependency, there is no test to verify it. In order for it to work, Quadlet needs to make sure the keys set by the user are placed after the ones it sets on its own. While this might be the case now, without a test, there is no guaranty that this behavior does not break in the future.

**Comment by jbtrystram** [comment]
@ygalblum I am looking at adding another test to verify that  `After=` removes the dependency,  but I can't see any assertion helper to use, so I am trying to add one. 

I started to add `assert-key-is-last`  which would call `UnitFile.LookupLast` but then I need the key to have no arguments.
The assertion then would be `assert-key-is-last-and-empty` but then it becomes awfully specific ! 
Do you have any other idea ? It's my first contrib to podman so I don't know my way around the codebase :)

**Comment by ygalblum** [comment]
@jbtrystram how about `assert-last-key-is-regex` and pass `^$`?

**Comment by jbtrystram** [comment]
@ygalblum neat suggestion, thanks ! 
I added that. I am having synchronisation issues when running `make test` locally though, let's see if the CI runs the test !

File: test/e2e/quadlet_test.go
Comment (ygalblum) [inline test/e2e/quadlet_test.go:186]: There should be a single expected value not a list

**Comment by rhatdan** [comment]
@jbtrystram What is going on with this PR?

**Comment by jbtrystram** [comment]
@rhatdan This is something i am working on my spare time and i hadn't had a lot of that lately, sorry. 
I'll follow up soon

**Comment by jbtrystram** [comment]
packit fails with : `packit.exceptions.PackitSRPMException: Preparation of the repository for creation of an SRPM failed: Reference at 'HEAD' does not exist`

**Comment by jbtrystram** [comment]
@ygalblum I picked that up and fixed it. The `quadlet.go` code behave as expected, as `quadlet --dryrun` on `unit-after-override.image` gives : 
```
---test.service---
## assert-last-key-is-regex "Unit" "After" "^$"

[Unit]
Wants=network-online.target
After=network-online.target
After=
SourcePath=/etc/containers/systemd/test.container
RequiresMountsFor=%t/containers

.....
```
However the test code with that `assert-last-key-is-regex` does not work yet. I'll keep working on it but still pushed, if you have some suggestions in the meanwhile i'll take them :)

File: test/e2e/quadlet_test.go
Comment (ygalblum) [inline test/e2e/quadlet_test.go:190]: I've found the error.
```suggestion
	matched, err := regexp.MatchString(regex, value)
	if err != nil || !matched {
		return false
	}
	return true
```

**Comment by ygalblum** [comment]
Another two tests are failing since they already check the `After` key. The check `assert-key-is` expects all the values in the correct order. So you need to change the following:
`test/e2e/quadlet/mount.container`, line 13:
```
## assert-key-is "Unit" "After" "network-online.target" "vol2-volume.service"
```

`test/e2e/quadlet/network.quadlet.container`, line 3:
```
## assert-key-is "Unit" "After" "network-online.target" "basic-network.service"
```

**Comment by jbtrystram** [comment]
Thanks @ygalblum for the help. 
I squashed the commits and rebased. The failing tests seem unrelated

**Review by ygalblum — APPROVED** [review]

**Comment by jbtrystram** [comment]
@ygalblum I rebased this to see if that help with the tests.

**Comment by ygalblum** [comment]
[test_fixes.txt](https://github.com/containers/podman/files/15399345/test_fixes.txt)
@jbtrystram The tests fail because they were checking the values of the `After` key and this PR changes them (as it automatically adds `network-online.target`.
I've attached a patch file you can use to fix the tests. When in the Podman base directory run:
```
patch -p1 < test_fixes.txt
```

**Comment by jbtrystram** [comment]
@ygalblum I got confused by your patch because I am 100% sure I already fixed those tests earlier. I think I messed up the rebase yesterday and pushed an out of date branch

**Comment by jbtrystram** [comment]
yeah, that's better. I found my previous work on another machine. 
Sorry for the trouble @ygalblum 

The test failure is related to buildah : 
```
  Error running buildah bud tests under podman.
         
         Is it possible that your PR breaks podman build in some way? Please
         review the test failure and double-check your changes.
         
         
         Please see test/buildah-bud/README.md for advice
```

**Review by ygalblum — APPROVED** [review]

**Comment by jbtrystram** [comment]
I don't think this failure have anything to do with my PR :
```
# podman run --name o8zHCMmnW1 --health-cmd=touch /terminate --sdnotify=healthy quay.io/libpod/testimage:20240123 sh -c while test \! -e /terminate; do sleep 0.1; done; echo finished
<+574ms> # finished
         # Error: container is stopped
<+004ms> # [ rc=126 (** EXPECTED 0 **) ]
         # #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
         # #| FAIL: exit code is 126; expected 0
         # #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by ygalblum** [comment]
@@containers/podman-maintainers PTAL

**Comment by mheon** [comment]
Known flake, restarted

**Comment by jbtrystram** [comment]
everything passed ! Can I get a LGTM ?

**Comment by mheon** [comment]
LGTM. This is probably good for inclusion in 5.1

**Pull request merged**
